metricbeat/module/prometheus/collector/collector.go

```diff
@@ -126,7 +128,10 @@ func MetricSetBuilder(namespace string, genFactory PromEventsGeneratorFactory) f
 
 // Fetch fetches data and reports it
 func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
-	m.once.Do(m.promEventsGen.Start)
+	m.once.Do(func() {
+		m.promEventsGen.Start()
+		m.eventGenStarted = true
+	})
```

Comment:
Fetch is never called concurrently (at least we are assuming this so far), so here we wouldn't need `m.once`.
```suggestion
	if !m.eventGenStarted {
		m.promEventsGen.Start()
		m.eventGenStarted = true
	}
```